Let users use variables as coordinates to a function in an auxkernel

GiudGiud opened the pull request:
I have a project where we have function data that is hardly convertible to another format (eg we would need to write an object that does what the function reading the csv does)
and an auxiliary variable that is spatial but not x y z (it s a distance to a wall). So it would fit great as a function argument conceptually. Unfortunately, ParsedFunction does not accept auxvariables.

So our two options (aside from this) would be:
- create a new matprop or auxkernel that includes the function work, and either includes the wall distance computation work or accept an aux variable. Keep this object in an app since it cant go in MOOSE
- create a new matprop or auxkernel that takes the function and the auxvariable as arguments and does this operation we have here but with dedicated names and only supports the 1 need we have. Also cant go in moose and it s just a less general version of this object

The concerns with having functions depend on variables:
- no automated dependency resolution that ensures the auxvariables are up to date when they are used in the function as arguments. Some of the functors (PPs) are not the best at this anyway
- blurs the lines between systems. Functions should be xyzt, AuxVariables kind of an output tool and material properties for having actual variable dependencies in terms that come in the equations
- loses derivatives. But that s the case with ALL auxkernels so that concern is not new

refs https://github.com/idaholab/moose/issues/26784

@youyeonc

GiudGiud committed:
Add an auxkernel to abuse functions into accepting variable (functor) arguments refs #26784

GiudGiud committed:
Add test and doco for functor abuse aux refs #26784

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2117324/) on b39bd07 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/27081/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2117368/) on b39bd07 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>9873be</th>
      <th colspan=4>#27081 b39bd0</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.30%</td>
      <td>85.30%</td>
      <td>+0.00%</td>
      <td>95.12%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>101214</td>
      <td>101254</td>
      <td>+40</td>
      <td>39</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>17444</td>
      <td>17445</td>
      <td>+1</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27081/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27081/coverage/framework)


## Modules coverage

### Fluid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>9873be</th>
      <th colspan=4>#27081 b39bd0</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.46%</td>
      <td>85.47%</td>
      <td>+0.01%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>7257</td>
      <td>7258</td>
      <td>+1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1235</td>
      <td>1234</td>
      <td>-1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27081/coverage_diff/fluid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27081/coverage/fluid_properties)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/27081/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented:
Fixed the pr title. 
Thoughts for the name?

FunctionWithFunctorCoordinatesAux?
FunctorPositionFunctionAux?

GiudGiud commented:
@dschwen thoughts on picking a new name?

GiudGiud commented:
Now that you are back @dschwen thoughts for the name?

dschwen commented:
CoupledVariableCoordinateFunctorAux?

GiudGiud commented:
It s more of a FunctionAux since the Function is used for the evaluations
and the variables are not coordinates, I use functor for those

so maybe FunctorCoordinatesFunctionAux ?

GiudGiud committed:
Address Daniel's review
rename to FunctorCoordinatesFunctionAux

GiudGiud commented:
Failures are unrelated

GiudGiud commented:
@lindsayad @dschwen

lindsayad commented on framework/src/auxkernels/FunctorCoordinatesFunctionAux.C:
```diff
@@ -0,0 +1,80 @@
+    _factor(getFunctor<Real>("factor")),
+    _is_fe(dynamic_cast<MooseVariableFE<Real> *>(&_var))
+{
+  if (!_is_fe && !dynamic_cast<MooseVariableFV<Real> *>(&_var))
```
This should be updated with the new linear finite volume variables. They support functor evaluation

GiudGiud replied:
addressed

GiudGiud replied:
it s not really about evaluation here though, it s more about "what can be set by an auxkernel" I dont know that we need this check even since we should complain upstream about an auxkernel being used on a non-real valued variable

lindsayad commented on framework/doc/content/source/auxkernels/FunctorCoordinatesFunctionAux.md:
```diff
@@ -0,0 +1,31 @@
+would be incomplete.
+
+!alert warning
+If using auxiliary variables as arguments to the `FunctorCoordinatesFunctionAux`, you must check that the ordering of execution
+is as expected, in that the auxiliary kernels that set those auxiliary variables are executed BEFORE the `FunctorCoordinatesFunctionAux`.
+This can be done with the [!param](/Debug/show_execution_order) parameter in the `[Debug]` block.
```
It's not enough that one aux kernel supplying an aux var needed for this new aux kernel is executed before within a given `execute_on`. It must be executed at some previous `execute_on`. This is because the aux variable functor evaluation reads from the `_sys.currentSolution` vector which, despite its name, is not updated until after aux kernels have been evaluated for a given `execute_on`. This is (somewhat) documented in https://github.com/idaholab/moose/issues/19396

GiudGiud replied:
I ll add a note. it s a shame to have caveats in the functor APIs which we wanted to solve all this

lindsayad replied:
We could probably fix 99% of use cases by drawing from the parallel solution if the variable kind is auxiliary. This will be wrong if someone tries to, for instance, fill a nodal finite element variable on a processor boundary with a nodal functor evaluation of an auxiliary finite volume variable (non-local values would be wrong in this case)

lindsayad replied:
Fixing the aux kernel limitations hasn't really been a priority since functor-driven simulations such as finite volume can almost always just use a functor material property in any place an aux variable might be used. But since most finite element simulation objects index at quadrature points, being able to fill aux variables instead of using a functor material property is a worthwhile endeavor

lindsayad replied:
> it s a shame to have caveats in the functor APIs which we wanted to solve all this

Indeed. However, calculations involving functor material properties and nonlinear variables, which was all I was initially targeting with the functor concept, are still caveat free

lindsayad reviewed (commented):
I actually very much like this PR

GiudGiud committed:
Address Alex's review
- allow for linear finite volume variables
- add note about auxkernel execution and variable updates

GiudGiud commented:
Test failures unrelated.
thanks for the review!
